### ISSUE TITLE
containers get scheduled on --reserved-cpus

### ISSUE THREAD

**Issue by iancoolidge** [kubernetes/kubernetes#115994]
### What happened?

Node has 64 vCPUs. CPU manager policy is 'static'

8 vCPUs are marked as resreved (via --reserved-cpus)
The remaining 56 vCPUs are all isolated via kernel command line.
systemd.cpu_affinity is set for all of the --reserved-cpus also (I think this is redundant with isolated CPU command line though)

Workload has two types of pods: those that request full integer CPUs (we'll call these 'dedicated CPU tasks') and those that do not. 'normal tasks'

result:
dedicated CPU tasks get correctly scheduled on the isolated vCPUs
normal tasks do not, and instead, get scheuled to the --reserved-cpus set.

### What did you expect to happen?

Both dedicated CPU tasks and normal tasks should all get scheduled on the 56 isolated vCPUs.

### How can we reproduce it (as minimally and precisely as possible)?

Based on feedback, if the bug is not obvious, I can try to build a minimized reproduction.

### Kubernetes version

Kubernetes v1.23.5-gke.1505

### Cloud provider

GKE-like environment (GDC Edge)

### OS version

Linux 5.15.71

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

containerd github.com/containerd/containerd v1.5.13-gke.0 a17ec496a95e55601607ca50828147e8ccaeebf1

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by xujyan** [comment]
> normal tasks do not, and instead, get scheuled to the --reserved-cpus set.

@iancoolidge did you mean the "8 vCPUs are marked as resreved" in your example?

What I observed is that these "normal tasks" get scheduled onto the full CPU set. (what would equivalent to 0-63 in your example), could that be what you mean instead?

**Comment by iancoolidge** [comment]
> > normal tasks do not, and instead, get scheuled to the --reserved-cpus set.
> 
> @iancoolidge did you mean the "8 vCPUs are marked as resreved" in your example?
> 
> What I observed is that these "normal tasks" get scheduled onto the full (including the reserved) CPU set (which would be equivalent to 0-63 in your example), could that be what you mean instead?

I did not see any of the 'normal' tasks getting assigned to any isolated vCPUs, I wonder if that is a function of loading though.

**Comment by SergeyKanzhelev** [comment]
/triage needs-information

@iancoolidge can you provide more information to make it easier to reproduce? Like pod definitions and such?

Also can you try it on the newer version that is still in support? 1.23 is out of support in OSS project.

**Comment by ffromani** [comment]
I managed to get some time to look at this. I think this it still applies for modern k8s. But I'd rather doublecheck in the (ancient) cpumanager implementation history if this was deliberate or just an accident which slipped through cracks.

When cpumanager manages non-exclusive cpus, the cpuset is unrestricted, because all the allocation logic sits in the `static` policy implementation, the `none` policy is basically all NOPs.

A possible fix could be always remove the restricted cpu set from the available cpu set for both policies for all pods regardless of the QoS, before any other allocation logic (ofc this applies only for exclusive cpus assignment, otherwise there's no further logic).

I can't help but wonder if this behavior should be gated by a cpumanager policy option or a feature gate, or if it can be considered just a bug. It's arguably a behavior which was there since long time, and people may have implicitely depend on it.

**Comment by ffromani** [comment]
cc @klueska

**Comment by iancoolidge** [comment]
@ffromani could you please provide some code links to what you are thinking of? I am trying to learn the codebase a little bit better

**Comment by ffromani** [comment]
> @ffromani could you please provide some code links to what you are thinking of? I am trying to learn the codebase a little bit better

I'd start from
https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.0/pkg/kubelet/cm/cpumanager/cpu_manager.go
https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.0/pkg/kubelet/cm/cpumanager/policy_options.go
https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.0/pkg/kubelet/cm/cpumanager/policy_none.go
https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.0/pkg/kubelet/cm/cpumanager/policy_static.go

and the related tests

**Comment by ffromani** [comment]
`NewManager` (https://github.com/kubernetes/kubernetes/blob/v1.28.0-alpha.0/pkg/kubelet/cm/cpumanager/cpu_manager.go#L154) knows the reserved CPUs set when given, or knows how to compute it when not given (and just given how many CPUs to reserve). It can pass this information to the `None` Policy like it does to the `Static` policy. The `None` Policy should carry around this set and can just substract from the set of available CPUs when allocating. I don't think there's need to add state management to the `None` policy, should be straightforward. Similarly, the `Static` policy should do the same when non-exclusive CPUs are needed.

So far so good. What's non-obvious if we can treat this as simple bug and just change with a release note or if we need to govern this behavior with a cpumanager policy option, a feature gate or both.

I'm inclined towards the latter, but let's see what the community thinks.

**Comment by klueska** [comment]
It's a bug. We can / should put the fix in the first patch release of 1.27 if it's ready by then.

**Comment by MarSik** [comment]
Kevin, I have known about this behavior for ages, but never considered it a bug honestly.

I always treated it as a way to let kernel scheduler do the right thing and keep the node stable. A case of quid for quo as the kernel is allowed to run interrupt handlers and other internal jobs on the k8s workload (for lack of a better name) cpus.

Simply said, I always expected burstable and best effort to not care where they run exactly as long as the cluster capacity limits are enforced properly.

So even though this might be a bug, it is one that actually might be beneficial for the node stability as it gives kernel some space to reshuffle the processes when it gets close to the system overload condition.


On the other hand, I often see people coming from other platforms to get confused by this when they look at the PSR value in ps and see container processes on the reserved cpus.

**Comment by MarSik** [comment]
@iancoolidge Btw, how are you isolating the 56 cores on the kernel command line?

Beware of isolcpus=domain,X-Y (the typical OS level partitioning kernel arg) as might not be doing what you want even if this issue is fixed. All burstable containers will end up on one of those 56 cores (the same or random depending on kernel version), but the processes will never move again or rebalance according to their cpu usage. The same will happen to threads started from those processes, they will all run together, because CPU balancing is disabled on those cores.

**Comment by ffromani** [comment]
@iancoolidge adding a new pool is something that sometimes pops out, but that would require a separate conversation, and likely a KEP. Regarding the fix for this bug, I understand you are interested and you would like to push a fix later on?

**Comment by iancoolidge** [comment]
@ffromani I am interested in getting a fix landed. If you can provide a fix, that's the best thing.

Do you have the bandwidth to support a fix?

**Comment by ffromani** [comment]
I have very limited bandwidth before kubecon EU, but merges are frozen up until April 11 anyway. I can help and I'll happy to review. Let's see what I can do.

**Comment by iancoolidge** [comment]
@MarSik One good reason for strict separation is enforcement of power savings policies on a per-CPU basis.

power savings optimizatoins can be aggressively set on --reserved-cpus, without materially impacting workload performance. There are tools emerging for configuring workload CPU power savings (cstate, pstate), but it's difficult to 'merge' those with a policy for --reserved-cpus.

**Comment by dastonzerg** [comment]
@ffromani Hi Francesco, this is Weipeng who communicated through Email with you yesterday : ) As per thread let's continue the discussion for the fix here. Others feel free to jump in.

I think the problem is with this line: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L204
I tried changing them to:
```
nonReservedCPUs := allCPUs.Difference(p.reservedCPUs)
s.SetDefaultCPUSet(nonReservedCPUs)
```
This seems to do what is expected during a test in our lab machine, which makes containers not in the "integer CPU requests within guaranteed pods" group to be allocated to the CPU set that excludes reserved CPUs.

You mentioned that this change might only work in a very limited set of cases as it is applied too late in the flow. Would like to know the right place to look at instead, thx!

**Comment by MarSik** [comment]
@iancoolidge That is a good point, but I guess we took a different route in OCP and cri-o. We actually configure the system for power saving as a whole and then disable c-states only when a guaranteed pod requests that. This allows us to save power on the unused workload cpus as well. See for example https://www.redhat.com/architect/power-management-openshift-edge

**Comment by iancoolidge** [comment]
@MarSik that is a neat implementation. However, it doesn't really address the problem that I pointed out. system CPUs and management things probably do tolerate aggressive cstate/pstate settings, but are prevented to run at those states if a pod that requests to disable c-states gets scheduled on them. It seems safe to assume that those will always be happening on exclusive CPUs, so that itself is probably not a big issue, but, wouldn't it be cleaner and more general to be able to specify the policies indendently? IMO, yes.

Another point to bring up is schedulable resources. When you provide --reserved-cpus, kubelet subtracts those CPUs from the 'schedulable resources' of the node. So, if you want to accurately represent system overhead, you need to specify a small --reserved-cpus, however, if all remaining CPUs are isolated, that causes non-guaranteed workloads AND system overhead to run on the same small number of CPUs, causing performance bottlenecking. You can isolate less CPUs, but then there is no guarantee that a guaranteed workload will be scheduled on an isolated CPU. If you isolate less CPus, and then set --reserved-cpus higher to compensate, then you are under-specifying schedulable resources on the system.

**Comment by ffromani** [comment]
> @ffromani Hi Francesco, this is Weipeng who communicated through Email with you yesterday : ) As per thread let's continue the discussion for the fix here. Others feel free to jump in.
> 
> I think the problem is with this line: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L204 I tried changing them to:
> 
> ```
> nonReservedCPUs := allCPUs.Difference(p.reservedCPUs)
> s.SetDefaultCPUSet(nonReservedCPUs)
> ```
> 
> This seems to do what is expected during a test in our lab machine, which makes containers not in the "integer CPU requests within guaranteed pods" group to be allocated to the CPU set that excludes reserved CPUs.
> 
> You mentioned that this change might only work in a very limited set of cases as it is applied too late in the flow. Would like to know the right place to look at instead, thx!

Hi! The very core idea here is correct.  Cases we should verify and cover I can think of on top of my head are:

1. is the recovery flow (reading back state from cpu_manager_state) handled correctly?
2. do we need changes to the none policy?
3. I'm wary in fixing the available cpu set inside `validateState`. This seems too implicit. We should try hard to make the exclusion of cpu set more visible and possibly rearrange the code to do this in `NewManager`.

**Comment by ffromani** [comment]
@dastonzerg (cc @iancoolidge ) consensus is reached this deserves a change, so maybe let's move the implementation discussion on a PR. Chatting on sig-node channel on k8s slack is also an option.

**Comment by dastonzerg** [comment]
@ffromani Hi Francesco plz take a look at the PR thx https://github.com/kubernetes/kubernetes/pull/118021

**Comment by MarSik** [comment]
Ok, I was thinking about this a bit more and I think you can consider my objection to be void.

My argumentation path for why this is indeed a bug and not a change in behavior is as follows:
- kubelet is supposed to be OS agnostic and delegate the OS specific bits to the runtime
- reserved-cpus is meant to define an explicit CPU set for OS system daemons and kubernetes system daemons
- kubelet and cpu manager should not assume the OS will even allow the overlap of the kubepod and system cpu sets in the generic case

I was thinking about making this configurable, but that hits the same argumentation of being OS agnostic. There are options now to keep the old behavior (which I still think is better for generic Linux based node) via the runtime (like NRI hooks).

**Comment by swatisehgal** [comment]
> /triage needs-information
> @iancoolidge can you provide more information to make it easier to reproduce? Like pod definitions and such?
> 
> Also can you try it on the newer version that is still in support? 1.23 is out of support in OSS project.

@iancoolidge I see there were questions asked as part of the initial triage process ([above](https://github.com/kubernetes/kubernetes/issues/115994#issuecomment-1470476108)) that are still unanswered.  Can you please provide a reproducer for this and confirm that this issue persists in newer versions of Kubernetes? Once you have managed to reproduce this on supported version(s) of Kubernetes, please update the issue description. Thanks!

**Comment by iancoolidge** [comment]
Fundamentally, I think the issue here is that, when we're telling kubelet about partitioning the computer's CPUs, we can only provide --reserved-cpus to give 2 pools: sharedpool+housekeeping, and exclusivepool.
The gap is that we need the ability to express 3 pools: housekeeping, sharedpool, and exlusivepool.

This is typically not consequential with typical enterrpise setup, but when you start to talk about running on the real-time kernel, or when you want to have differentiated powersave policies for housekeeping vs sharedpool, this becomes important.

How are users currently separating housekeeping from sharedpool?

**Comment by MarSik** [comment]
We don't separate them explicitly atm. Capacity wise k8s makes sure the shared pool will not overflow to the housekeeping pool. And then we let the kernel decide. Our relatime processes are only running in the exclusive pool (cpu manager). The rest requires nothing special.

If anything, we used to have an issue with mixed shared + exclusive cpus due to nohz_full kernel arg being static configuration. But that has been solved by https://github.com/torvalds/linux/commit/88c56cfeaec4642aee8aac58b38d5708c6aae0d3

In other words, we do not care too much about the fact that the kernel can utilize the housekeeping cpus to run shared pool processes. There is enough space in the shared pool to not force the kernel do that, but when it wants/has to do that, it can.

**Comment by iancoolidge** [comment]
> Capacity wise k8s makes sure the shared pool will not overflow to the housekeeping pool. And then we let the kernel decide. 

Can you elaborate a bit please?

**Comment by MarSik** [comment]
Well, Allocatable is computed as Capacity - reserved. So even if you fill up the node capacity and all containers use all their allocated cpus, the capacity of the reserved cpus is not affected. Of course as long as you set limits as well.

**Comment by iancoolidge** [comment]
@MarSik a related question.

We use --reserved-cpus to express to kubelet which of our CPUs are isolated. (the non-reserved ones go into the exclusive pool)
This seems to reduce the allocatable node capacity to TOTAL_CPUS - RESERVED_CPUS, which seems to prevent workload deployment that the node definitely *has* capacity for.

Is there a better way to make kubelet agree on the isolated CPUs while not reducing the node capacity?

**Comment by MarSik** [comment]
@iancoolidge I think I finally understand what your use case is, but let me make sure I have it right:

You have workloads that are not managed by k8s and use dedicated cpus. That is why you mentioned isolcpus?. Those cpus are not for low latency container workloads, but for something else.

You then use the system-reserved-cpus to tell k8s about the cores that are off-limits, but the burstable pods can still run there (as expected, because k8s does not know about isolcpus).

Am I right?

**Comment by MarSik** [comment]
Ok, I reread the history a bit more carefully and it seems that no, you do not have the extra non-k8s workloads. It makes an interesting use case (that k8s probably does not support though).

> We use --reserved-cpus to express to kubelet which of our CPUs are isolated. 

Well.. but that is actually not what the option was designed for. It tells kubelet which cpus are "reserved" for the operating system: https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list

_"reserved-cpus is meant to define an explicit CPU set for OS system daemons and kubernetes system daemons."_

The rest are both exclusive + shared at the same time (= the pod capacity of the node) and the distinction is done at runtime depending on the pods running on the node by cpu manager.

So system reserved should really be small to accomodate the operating system only and not lower the capacity of the node too much.

I am OK with making the behavior tidier (by removing system reserved from the cpumask of burstable), but I think the argumentation for that needs to be consistent.

Btw, I just checked and our power management story is different too. Due to exec probes load and kernel networking, we are actually allowing higher frequencies for our system reserved cpus.

**Comment by iancoolidge** [comment]
> Well.. but that is actually not what the option was designed for. It tells kubelet which cpus are "reserved" for the operating system

To be totally clear, I am not expecting that any workkloads are running on --reserved-cpus, we are trying to use the *inverse* of --reserved-cpus to run our workloads.

> The rest are both exclusive + shared at the same time (= the pod capacity of the node) and the distinction is done at runtime depending on the pods running on the node by cpu manager.

Can you please elaborate on this?

My understanding is different - 

cpumanager is not smart enough to take the non --reserved-cpus and dynamically divide them between shared and exclusive, it does not even attempt this. it uses --reserved-cpus for shared pool, and inverse of --reserved-cpus for exclusive pool.

**Comment by jingczhang** [comment]
Reserved cores are not reserved have been causing real issues in production for years now, harms are on both sides (on both system processes and workloads). Various attempts have been made, like this one, to fix the issue but not getting anywhere. I have created a KEP for this, please review / comment / throw your support:
https://github.com/kubernetes/enhancements/pull/4541

**Comment by kad** [comment]
> Reserved cores are not reserved have been causing real issues in production for years now, harms are on both sides (on both system processes and workloads). Various attempts have been made, like this one, to fix the issue but not getting anywhere. I have created a KEP for this, please review / comment / throw your support: [kubernetes/enhancements#4541](https://github.com/kubernetes/enhancements/pull/4541)

Just a reminder: for situations where cpumanager is not smart enough and cluster require more fine-tuned solutions, since end of 2022 there is a way to achieve it via https://github.com/containerd/nri/ and https://github.com/containers/nri-plugins/

**Comment by jingczhang** [comment]
The balloons policy is mentioned in the https://github.com/kubernetes/enhancements/pull/4541, please take a look. The ballons policy requires additional software and tuning, and please correct me, it is not a straight-forward configuration to address this reserved core issue plus it is not compliant with the widely used Topology Manager/Memory Manager/CPU Manager.

**Comment by kad** [comment]
"Topology-aware" is the zero-configuration analog of "static" policy + all the functionality from rest of the managers + more (e.g. support for modern hardware, memory tiering, etc).

My point is about other thing, however: if you have specific workload requirements, you don't necessarily need to patch kubelet, create policy that suits your particular customer needs. There are different examples on how to do that and build on top of it.

**Comment by RiskyAdventure** [comment]
Thank you for opening this issue. I'd like to outline my use case to see if it fits a wider audience of why I think this feature would be so helpful. 

I run CPU policy static for spark workloads across 100's of nodes. Occasionally, during the startup process, I will over utilize a particular node for a very short time frame. This has the unfortunate effect of disrupting the NAPI processing and I get a runaway condition where rx networking traffic errors begin and cause a cascading problem on the node. 

What I was hoping reserved CPU would do, is isolate a set of CPU on these large 96 core boxes, so things like networking processes, PID 1, etc. will not be interrupted by sidecars or anything workload related. My hope is that this would allow me to compress the workload CPU for a short amount of time safely, which does not seem possible today.

**Comment by ffromani** [comment]
https://github.com/kubernetes/enhancements/issues/4540 should fix this issue - feature currently in alpha

**Comment by swatisehgal** [comment]
> [kubernetes/enhancements#4540](https://github.com/kubernetes/enhancements/issues/4540) should fix this issue - feature currently in alpha

Yes, the enhancement issue is open to track the progress of the feature so let's close this issue and the associated PR.

/close
